backend/agentd/session.go

```diff
@@ -427,7 +430,28 @@ func (s *Session) stop() {
 		}
 	}()
 
+	// Send a close message to ensure the agent closes its connection if the
+	// connection is not already closed
+	if !s.conn.Closed() {
+		if err := s.conn.SendCloseMessage(); err != nil {
+			logger.WithError(err).Warning("unexpected error while sending a close message to the agent")
+		}
+	}
+
 	sessionCounter.WithLabelValues(s.cfg.Namespace).Dec()
+
+	// Gracefully wait for the send and receiver to exit, but force the websocket
+	// connection to close itself after the grace period
+	done := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(closeGracePeriod):
+	}
+
 	s.wg.Wait()
```

Comment:
This `Wait()` is redundant now

Reply:
Ah yes good catch, I thought I removed it 🤦